### PR TITLE
Add basic primitives for lightning port

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #

## Proposed Changes
- added lightning module with basic initialization options
- refactored scheduler, optimizer, and metric creational functions
- added unit tests for the newly created primitives
- added new unit tests to CI pipeline

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

File: GANDLF/losses/loss_calculators.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        self.loss = get_loss(self.params)
+
+    @abstractmethod
+    def __call__(
+        self, prediction: torch.Tensor, target: torch.Tensor, *args
+    ) -> torch.Tensor:
+        pass
+
+
+class LossCalculatorStdnet(AbstractLossCalculator):
```
[the remark this answers is not in the document]
Comment (sarthakpati) [reply GANDLF/losses/loss_calculators.py]: I think this should be `SDNet`...?

File: GANDLF/losses/loss_calculators.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    def __init__(self, params):
+        super().__init__(params)
+        self.l1_loss = get_loss(params)
+        self.kld_loss = get_loss(params)
+        self.mse_loss = get_loss(params)
```
Comment (sarthakpati) [inline GANDLF/losses/loss_calculators.py]: How are the different losses triaged in the config?

File: GANDLF/losses/loss_calculators.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            return self.loss(prediction, target, self.params)
+
+
+class LossCalculatorDeepSupervision(AbstractLossCalculator):
+    def __init__(self, params):
+        super().__init__(params)
+        # This was taken from current Gandlf code, but I am not sure if
+        # we should have this set rigidly here, as it enforces the number of
+        # classes to be 4.
+        self.loss_weights = [0.5, 0.25, 0.175, 0.075]
```
Comment (sarthakpati) [inline GANDLF/losses/loss_calculators.py]: I agree to make this more flexible. The GaNDLF code was specifically designed for the 4-layer DeepUNet, and no longer makes sense if we can have multi-layer DeepUNet working.

File: GANDLF/metrics/metric_calculators.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            return metric_value.tolist()
+
+    @abstractmethod
+    def __call__(
+        self, prediction: torch.Tensor, target: torch.Tensor, *args
+    ) -> torch.Tensor:
+        pass
+
+
+class MetricCalculatorStdnet(AbstractMetricCalculator):
```
Comment (sarthakpati) [inline GANDLF/metrics/metric_calculators.py]: Same here (it should be `SDNet`, perhaps?).

File: testing/test_lightning_components.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+from GANDLF.parseConfig import parseConfig
+from GANDLF.utils.write_parse import parseTrainingCSV
+from GANDLF.utils import populate_header_in_parameters
+
+testingDir = Path(__file__).parent.absolute().__str__()
+
+
+def add_mock_config_params(config):
+    config["penalty_weights"] = [0.5, 0.25, 0.175, 0.075]
+    config["model"]["class_list"] = [0, 1, 2, 3]
```
Comment (sarthakpati) [inline testing/test_lightning_components.py]: This can be changed based on the support for multi-layer deep supervision.

File: testing/test_lightning_components.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    dummy_target = torch.randint(0, 4, (4, 4, 4, 4))
+    metric = metric_calculator(dummy_preds, dummy_target)
+    for metric, value in metric.items():
+        assert not math.isnan(value), f"Metric {metric} has NaN values"
+
+
+#### LIGHTNING MODULE ####
+
+
+def test_port_model_initalization():
```
[the remark this answers is not in the document]
Comment (sarthakpati) [reply testing/test_lightning_components.py]: `initialization`?

File: GANDLF/losses/loss_calculators.py (hunk 2 above)
Comment (szmazurek) [reply GANDLF/losses/loss_calculators.py]: Will be resolved in another PR

File: GANDLF/metrics/metric_calculators.py (hunk 4 above)
Comment (szmazurek) [reply GANDLF/metrics/metric_calculators.py]: yes

File: GANDLF/losses/loss_calculators.py (hunk 3 above)
Comment (szmazurek) [reply GANDLF/losses/loss_calculators.py]: How would we do that? Should it be a configurable parameter? If so, what is the default (if we decide to use default)?

File: testing/test_lightning_components.py (hunk 5 above)
Comment (szmazurek) [reply testing/test_lightning_components.py]: Could you explain/point me towards the part of docs saying more? Not sure if I fully get that and what should be specified there.

**Comment by sarthakpati** [comment]
Do you need any feedback on the comments earlier or (since this PR has been closed) are you good?
